### PR TITLE
[Feature] Support Flink 1.16 submit

### LINKED ISSUE #1620: [Feature] Support Flink 1.16 submit
### Search before asking

- [X] I had searched in the [feature](https://github.com/apache/streampark/issues?q=is%3Aissue+label%3A%22Feature%22) and found no similar feature requirement.

### Description

I have noticed that flink 1.16 has recently created release-rc0[1], and trust that the official release version will be coming soon. At that time, streampark will be support used flink 1.16 to submit job.

For more information about Flink release 1.16, can refer to
[1]https://github.com/apache/flink/releases/tag/release-1.16.0-rc0
[2]https://cwiki.apache.org/confluence/display/FLINK/1.16+Release

### Usage Scenario

Support used flink 1.16 to submit job.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by lvshaokang**
## What changes were proposed in this pull request

Issue Number: close #1620  

Support used flink 1.16 to submit job. 

## Brief change log

Add 1.16 shims

## Verifying this change

This change added tests and can be verified as follows:

- Manually verified submit job using yarn with custom code/sql by testing locally.
- Verifyed using jar syntax checking.

## Does this pull request potentially affect one of the following parts
 - Dependencies (does it add or upgrade a dependency): yes

### CODE REVIEW

**Comment by lvshaokang** [comment]
cc @wolfboys @1996fanrui

File: streampark-flink/streampark-flink-proxy/src/main/scala/org/apache/streampark/flink/proxy/FlinkShimsProxy.scala [hunk 1]
```diff
@@ -36,7 +36,7 @@ object FlinkShimsProxy extends Logger {
   )
 
   private[this] val SHIMS_PATTERN = Pattern.compile(
-    "streampark-flink-shims_flink-(1.12|1.13|1.14|1.15)_(2.11|2.12)-(.*).jar",
+    "streampark-flink-shims_flink-(1.12|1.13|1.14|1.15|1.16)_(2.11|2.12)-(.*).jar",
```
Comment (1996fanrui) [inline streampark-flink/streampark-flink-proxy/src/main/scala/org/apache/streampark/flink/proxy/FlinkShimsProxy.scala:39]: Hi @lvshaokang , thanks for your contribution. Most look good.

Too many pattern, please extract a const field.

**Comment by wolfboys** [comment]
hi lvshaokang:

 thanks for your contribution. I see the shims code for flink-1.16 is exactly the same as flink-1.15, So I don't think it's necessary for us to implement flink-1.16, If the flink version of the user's job is flink-1.16, we can select shims-1.15 for the corresponding shims version in the program, This is just my assumption, you can actually verify whether it is feasible

**Comment by lvshaokang** [comment]
> So I don't think it's necessary for us to implement flink-1.16, If the flink version of the user's job is flink-1.16, we can select shims-1.15 for the corresponding shims version in the program

@wolfboys , In my test step, i had some problems.
1. I remove all same codes and use the shims 1.15 dependency, but the mvn package jar without any shims 1.15 class
2. So, I modified the maven shade plugin to this,
```
<artifactSet>
  <includes>
    <include>org.apache.flink:flink-table-api-scala-bridge_${scala.binary.version}</include>
    <include>org.apache.streampark:streampark-flink-shims_flink-1.15_${scala.binary.version}</include>
 </includes>
</artifactSet>
```
I found the jar file contains the class, but i submit job, it throws a exception following this:
it used the StreamTableEnvironment of 1.15, rather then 1.16
```
Caused by: java.lang.NoSuchMethodError: org.apache.flink.table.catalog.FunctionCatalog.<init>(Lorg/apache/flink/configuration/ReadableConfig;Lorg/apache/flink/table/catalog/CatalogManager;Lorg/apache/flink/table/module/ModuleManager;)V
	at org.apache.flink.table.api.bridge.scala.internal.StreamTableEnvironmentImpl$.create(StreamTableEnvironmentImpl.scala:324) ~[streampark-flink-shims_flink-1.16_2.12-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
	at org.apache.flink.table.api.bridge.scala.StreamTableEnvironment$.create(StreamTableEnvironment.scala:899) ~[streampark-flink-shims_flink-1.16_2.12-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
	at org.apache.streampark.flink.core.FlinkTableInitializer.initEnvironment(FlinkTableInitializer.scala:247) ~[streampark-flink-sqlclient_2.12-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
	at org.apache.streampark.flink.core.FlinkTableInitializer$.initialize(FlinkTableInitializer.scala:78) ~[streampark-flink-sqlclient_2.12-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
	at org.apache.streampark.flink.core.scala.FlinkStreamTable.init(FlinkStreamTable.scala:54) ~[streampark-flink-sqlclient_2.12-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
	at org.apache.streampark.flink.core.scala.FlinkStreamTable.main(FlinkStreamTable.scala:45) ~[streampark-flink-sqlclient_2.12-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
	at org.apache.streampark.flink.core.scala.FlinkStreamTable.main$(FlinkStreamTable.scala:44) ~[streampark-flink-sqlclient_2.12-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
	at org.apache.streampark.flink.cli.SqlClient$StreamSqlApp$.main(SqlClient.scala:63) ~[streampark-flink-sqlclient_2.12-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
	at org.apache.streampark.flink.cli.SqlClient$.delayedEndpoint$org$apache$streampark$flink$cli$SqlClient$1(SqlClient.scala:55) ~[streampark-flink-sqlclient_2.12-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
	at org.apache.streampark.flink.cli.SqlClient$delayedInit$body.apply(SqlClient.scala:30) ~[streampark-flink-sqlclient_2.12-1.2.4-SNAPSHOT.jar:1.2.4-SNAPSHOT]
	at scala.Function0.apply$mcV$sp(Function0.scala:34) ~[flink-scala_2.12-1.16.0.jar:1.16.0]
	at scala.Function0.apply$mcV$sp$(Function0.scala:34) ~[flink-scala_2.12-1.16.0.jar:1.16.0]
```
Meanwhile,I found that the jar package size is a bit larger compared to the previous implementation. 
before: 57927
after: 60217

File: streampark-flink/streampark-flink-proxy/src/main/scala/org/apache/streampark/flink/proxy/FlinkShimsProxy.scala (hunk 1 above)
Comment (lvshaokang) [reply streampark-flink/streampark-flink-proxy/src/main/scala/org/apache/streampark/flink/proxy/FlinkShimsProxy.scala:39]: > Too many pattern, please extract a const field.

Hi, @1996fanrui . What you mean is that extract regex pattern to a const field?
such this `streampark-flink-shims_flink-(1.12|1.13|1.14|1.15|1.16)_(2.11|2.12)-(.*).jar, (^|.*)streampark-flink-shims_flink-(1.12|1.13|1.14|1.15|1.16)-(.*).jar$`

**Review by wolfboys — APPROVED** [review]
LGTM

It can indeed pass the test, but the adaptation codes of flink-1.15 and flink-1.16 are exactly the same. This part can be improved and optimized, but I have tried it, and it is not so simple. I can continue to follow up and improve this part of the code. Merged first, in the next few PRs, I will work on this.

**Pull request merged**